backend/api/v1/instance_service.go

```diff
@@ -1035,14 +1041,98 @@ func (s *InstanceService) convertToDataSourceMessages(dataSources []*v1pb.DataSo
+	if externalSecret == nil {
+		return nil, nil
+	}
+	secret := new(v1pb.DataSourceExternalSecret)
+	if err := convertV1PbToStorePb(externalSecret, secret); err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to convert external secret with error: %v", err.Error())
+	}
+
+	resp := &v1pb.DataSourceExternalSecret{
+		SecretType: secret.SecretType,
+		Url:        secret.Url,
+		AuthType:   secret.AuthType,
+		EngineName: secret.EngineName,
+		SecretName: secret.SecretName,
+		KeyName:    secret.KeyName,
+	}
+
+	// clear sensitive data.
+	switch resp.AuthType {
+	case v1pb.DataSourceExternalSecret_APP_ROLE:
+		resp.AuthOption = &v1pb.DataSourceExternalSecret_AppRole{
+			AppRole: &v1pb.DataSourceExternalSecret_AppRoleAuthOption{},
+		}
+	case v1pb.DataSourceExternalSecret_TOKEN:
+		resp.AuthOption = &v1pb.DataSourceExternalSecret_Token{
+			Token: "",
+		}
+	}
+
+	return resp, nil
+}
+
+func convertToV1DataSources(dataSources []*store.DataSourceMessage) ([]*v1pb.DataSource, error) {
+	dataSourceList := []*v1pb.DataSource{}
+	for _, ds := range dataSources {
+		externalSecret, err := convertToV1DataSourceExternalSecret(ds.ExternalSecret)
+		if err != nil {
+			return nil, err
+		}
+
+		dataSourceType := v1pb.DataSourceType_DATA_SOURCE_UNSPECIFIED
+		switch ds.Type {
+		case api.Admin:
+			dataSourceType = v1pb.DataSourceType_ADMIN
+		case api.RO:
+			dataSourceType = v1pb.DataSourceType_READ_ONLY
+		}
+
+		dataSourceList = append(dataSourceList, &v1pb.DataSource{
+			Id:       ds.ID,
+			Type:     dataSourceType,
+			Username: ds.Username,
+			// We don't return the password and SSLs on reads.
+			Host:                   ds.Host,
+			Port:                   ds.Port,
+			Database:               ds.Database,
+			Srv:                    ds.SRV,
+			AuthenticationDatabase: ds.AuthenticationDatabase,
+			Sid:                    ds.SID,
+			ServiceName:            ds.ServiceName,
+			ExternalSecret:         externalSecret,
+		})
+	}
+
+	return dataSourceList, nil
+}
+
+func convertToStoreDataSourceExternalSecret(externalSecret *v1pb.DataSourceExternalSecret) (*storepb.DataSourceExternalSecret, error) {
+	if externalSecret == nil {
+		return nil, nil
+	}
+	secret := new(storepb.DataSourceExternalSecret)
+	if err := convertV1PbToStorePb(externalSecret, secret); err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to convert external secret with error: %v", err.Error())
+	}
+
+	return secret, nil
+}
+
 func (s *InstanceService) convertToDataSourceMessage(dataSource *v1pb.DataSource) (*store.DataSourceMessage, error) {
 	dsType, err := convertDataSourceTp(dataSource.Type)
 	if err != nil {
 		return nil, err
 	}
+	externalSecret, err := convertToStoreDataSourceExternalSecret(dataSource.ExternalSecret)
+	if err != nil {
+		return nil, err
+	}
 
 	return &store.DataSourceMessage{
-		ID:                                 dataSource.Id,
+		ID:                                 uuid.NewString(),
```

Comment:
Might still be better to let caller/client/frontend pass in the id where (data source) id is the declarative identifier.

Reply:
Reverted. But seems we're missing unique check for the data source id, I can add this in next PR